[bitnami/kibana] remove unsable function `kibana_set_key_value` function and move to `kibana_conf_set`

chocolattej opened the pull request:
### Description of the change

remove unsable function `kibana_set_key_value` function and move to `kibana_conf_set` in `containers/bitnami/kibana/8/debian-12/rootfs/opt/bitnami/scripts/libkibana.sh`

### Possible drawbacks

- have to configure the `keystore-password` in kibana.yaml might expose security concern or somethings related.

### Applicable issues

<!-- Enter any applicable Issues here (You can reference an issue using #) -->
- fixes #76379

### Additional information

- this #76379 causing unable to use binami/kibana images with Java Keystore Configuration the full Issues report is in refer issue
- I have already tesing in my local environment and its work properly as expects
- keystore password configuration is exists on kibana.yaml
```yaml
path:
  data: /bitnami/kibana/data
pid:
  file: /opt/bitnami/kibana/tmp/kibana.pid
server:
  host: 0.0.0.0
  port: 5061
  ssl:
    enabled: true
    keystore:
      path: /opt/bitnami/kibana/config/certs/kibana.keystore.p12
      password: kibana-password
elasticsearch:
  hosts: https://elasticsearch:9200
  username: kibana_system
  password: kibana
  ssl:
    verificationMode: certificate
    truststore:
      path: /opt/bitnami/kibana/config/certs/elasticsearch.truststore.p12
      password: elasticsearch-password
```
and no Error logs exist on starting container process
- ENV in images build testing process
```.env
KIBANA_ELASTICSEARCH_URL: elasticsearch
KIBANA_ELASTICSEARCH_PORT_NUMBER: 9200
KIBANA_PASSWORD: kibana
KIBANA_PORT_NUMBER: 5061
KIBANA_SERVER_ENABLE_TLS: true
KIBANA_ELASTICSEARCH_ENABLE_TLS: true
KIBANA_ELASTICSEARCH_TLS_VERIFICATION_MODE: certificate
KIBANA_ELASTICSEARCH_TRUSTSTORE_LOCATION: /opt/bitnami/kibana/config/certs/elasticsearch.truststore.p12
KIBANA_ELASTICSEARCH_TRUSTSTORE_PASSWORD: elasticsearch-password
KIBANA_ELASTICSEARCH_TLS_USE_PEM: false
KIBANA_SERVER_KEYSTORE_LOCATION: /opt/bitnami/kibana/config/certs/kibana.keystore.p12
KIBANA_SERVER_KEYSTORE_PASSWORD: kibana-password
```

- I'm quite new here if there anythings unclear or need more information, feel free to reach me anytime.

chocolattej committed:
[bitnami/kibana] remove unsable function kibana_set_key_value function and move to kibana_conf_set in containers/bitnami/kibana/8/debian-12/rootfs/opt/bitnami/scripts/libkibana.sh

Signed-off-by: CHOCOLATTE <w.wiriya.m@gmail.com>

dgomezleon commented:
Hi @chocolattej ,

Thanks for contributing. 
As you can see [here](https://www.elastic.co/guide/en/kibana/current/secure-settings.html), `kibana-keystore` is used for sensitive string settings, like authentication credentials. I would not remove it.

Have you tried to fix the typo by only changing the code below?:

```
                    if [[ "$SERVER_FLAVOR" = "opensearch-dashboards" ]]; then
                        kibana_conf_set "server.ssl.keystore.password" "$SERVER_KEY_PASSWORD"
                    else
                        kibana_set_key_value "server.ssl.keystore.password" "$SERVER_KEY_PASSWORD"
                    fi
```

with

```
                    if [[ "$SERVER_FLAVOR" = "opensearch-dashboards" ]]; then
                        kibana_conf_set "server.ssl.keystore.password" "$SERVER_KEYSTORE_PASSWORD"
                    else
                        kibana_set_key_value "server.ssl.keystore.password" "$SERVER_KEYSTORE_PASSWORD"
                    fi
```


If that fixed the reported issue, I consider it a better option.

chocolattej commented:
Hi @dgomezleon ,
Thanks for your discuss.
You could I have already tried it in [detailed issues report](https://github.com/bitnami/containers/issues/76379#issuecomment-2606524625) 

along with this function
```bash
kibana_set_key_value() {
    local key="${1:?missing key}"
    local value="${2:?missing value}"

    debug "Storing key: ${key}"
    kibana-keystore add --stdin --force "$key" <<<"$value"
}
```

this is mannual re-produce the set_key_value_function in this line `kibana-keystore add --stdin --force "$key" <<<"$value"` and show result in
```bash
I have no name!@kibana-6c5bf7568c-fszkp:/$ kibana-keystore add --stdin --force "server.ssl.keystore.password" <<<"$SERVER_KEY_PASSWORD"
Kibana is currently running with legacy OpenSSL providers enabled! For details and instructions on how to disable see https://www.elastic.co/guide/en/kibana/8.17/production.html#openssl-legacy-provider
ERROR: Kibana keystore not found. Use 'create' command to create one.
```

and here is the log in starting container that show before container error and tried to restart
```
kibana 07:16:31.91 DEBUG ==> Storing key: server.ssl.keystore.password
Kibana is currently running with legacy OpenSSL providers enabled! For details and instructions on how to disable see https://www.elastic.co/guide/en/kibana/8.17/production.html#openssl-legacy-provider
ERROR: Kibana keystore not found. Use 'create' command to create one.
```

the `ERROR: Kibana keystore not found. Use 'create' command to create one.` message show in terminal is the same
so I think the problem is `set_key_value_function` not work properly as expected. I assume that `kibana-keystore cli` is could not found the key properties when we import key file instead of create with its own command and I do not know about it much, so my solution is remove this function that could not be use.

another solution that could be possible is find the way to import key file that `kibana-keystore cli` could find them so we could use set_key_value_function as expected

dgomezleon commented:
Hi @chocolattej ,

Thanks for the detailed explanation.
Have you considered this?: 

1. The error  `ERROR: Kibana keystore not found. Use 'create' command to create one.` could be solved by creating it previously with `kibana-keystore create`.
2. The warning `Kibana is currently running with legacy OpenSSL providers enabled! For details and instructions on how to disable see https://www.elastic.co/guide/en/kibana/8.17/production.html#openssl-legacy-provider` could be avoided by removing `--openssl-legacy-provider` from `/opt/bitnami/kibana/config/node.options` (OPTIONAL)

Could you give these a try and let me know if this solution work for you?

chocolattej commented:
Hi @dgomezleon  Sorry for late response. quite a tough week.
I do some research about `kibana-keystore` tools and imo its not reccommended mannually import until they update kibana-keystore tools instead, here is explanation

here is `bin/kibana-keystore`  code snippet

```bash
#!/bin/sh
SCRIPT=$0

# SCRIPT may be an arbitrarily deep series of symlinks. Loop until we have the concrete path.
while [ -h "$SCRIPT" ] ; do
  ls=$(ls -ld "$SCRIPT")
  # Drop everything prior to ->
  link=$(expr "$ls" : '.*-> \(.*\)$')
  if expr "$link" : '/.*' > /dev/null; then
    SCRIPT="$link"
  else
    SCRIPT=$(dirname "$SCRIPT")/"$link"
  fi
done

DIR="$(dirname "${SCRIPT}")/.."
CONFIG_DIR=${KBN_PATH_CONF:-"$DIR/config"}
NODE="${DIR}/node/glibc-217/bin/node"
test -x "$NODE"
if [ ! -x "$NODE" ]; then
  echo "unable to find usable node.js executable."
  exit 1
fi

if [ -f "${CONFIG_DIR}/node.options" ]; then
  KBN_NODE_OPTS="$(grep -v ^# < ${CONFIG_DIR}/node.options | xargs)"
fi

NODE_OPTIONS="$KBN_NODE_OPTS $NODE_OPTIONS" NODE_ENV=production "${NODE}" "${DIR}/src/cli_keystore/dist" "$@"
```

they using node.js to building up this tool with this `${DIR}/src/cli_keystore/...` directory
```bash
$ ls -l src/cli_keystore/
total 44
-rw-r--r-- 1 root root 2799 Jan 21 16:57 add.js
-rw-r--r-- 1 root root 2372 Feb 10 05:55 cli_keystore.js
-rw-r--r-- 1 root root 1601 Jan 21 16:57 create.js
-rw-r--r-- 1 root root  545 Jan 21 16:57 dist.js
-rw-r--r-- 1 root root 1396 Jan 21 16:57 has_passwd.js
-rw-r--r-- 1 root root 1139 Jan 21 16:57 list.js
-rw-r--r-- 1 root root 1344 Jan 21 16:57 passwd.js
-rw-r--r-- 1 root root  922 Jan 21 16:57 remove.js
-rw-r--r-- 1 root root 1978 Jan 21 16:57 show.js
-rw-r--r-- 1 root root  411 Jan 21 16:57 tsconfig.json
```

so I try to understanding create method to implement import module

```bash
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.create = create;
exports.createCli = createCli;
var _logger = require("../cli/logger");
var _utils = require("../cli/keystore/utils");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the "Elastic License
 * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
 * Public License v 1"; you may not use this file except in compliance with, at
 * your election, the "Elastic License 2.0", the "GNU Affero General Public
 * License v3.0 only", or the "Server Side Public License, v 1".
 */

async function create(keystore, options = {}) {
  const logger = new _logger.Logger(options);
  if (keystore.exists()) {
    const overwrite = await (0, _utils.confirm)('A Kibana keystore already exists. Overwrite?');
    if (!overwrite) {
      return logger.log('Exiting without modifying keystore.');
    }
  }
  keystore.reset();
  if (options.password) {
    const password = await (0, _utils.question)('Enter new password for the kibana keystore (empty for no password)', {
      mask: '*'
    });
    if (password) keystore.setPassword(password);
  }
  keystore.save();
  logger.log(`Created Kibana keystore in ${keystore.path}`);
}
function createCli(program, keystore) {
  program.command('create').description('Creates a new Kibana keystore').option('-p, --password', 'Prompt for password to encrypt the keystore').option('-s, --silent', 'Show minimal output').action(create.bind(null, keystore));
```

so I trace back to KeyStore class that they're  using in this `src/cli/keystore/keystore.js`

```js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Keystore = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _fs = require("fs");
var _crypto = require("crypto");
var _prompt = require("./utils/prompt");
var errors = _interopRequireWildcard(require("./errors"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the "Elastic License
 * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
 * Public License v 1"; you may not use this file except in compliance with, at
 * your election, the "Elastic License 2.0", the "GNU Affero General Public
 * License v3.0 only", or the "Server Side Public License, v 1".
 */

const VERSION = 1;
const ALGORITHM = 'aes-256-gcm';
const ITERATIONS = 10000;
class Keystore {
  static async initialize(path, password) {
    const keystore = new Keystore(path, password);
    await keystore.load();
    return keystore;
  }
  constructor(path, password = '') {
    this.path = path;
    this.password = password;
    this.reset();
  }
  static encrypt(text, password = '') {
    const iv = (0, _crypto.randomBytes)(12);
    const salt = (0, _crypto.randomBytes)(64);
    const key = (0, _crypto.pbkdf2Sync)(password, salt, ITERATIONS, 32, 'sha512');
    const cipher = (0, _crypto.createCipheriv)(ALGORITHM, key, iv);
    const ciphertext = Buffer.concat([cipher.update(text, 'utf8'), cipher.final()]);
    const tag = cipher.getAuthTag();
    return Buffer.concat([salt, iv, tag, ciphertext]).toString('base64');
  }
  static decrypt(data, password = '') {
    try {
      const bData = Buffer.from(data, 'base64');

      // convert data to buffers
      const salt = bData.slice(0, 64);
      const iv = bData.slice(64, 76);
      const tag = bData.slice(76, 92);
      const text = bData.slice(92);
      const key = (0, _crypto.pbkdf2Sync)(password, salt, ITERATIONS, 32, 'sha512');
      const decipher = (0, _crypto.createDecipheriv)(ALGORITHM, key, iv);
      decipher.setAuthTag(tag);
      return decipher.update(text, 'binary', 'utf8') + decipher.final('utf8');
    } catch (e) {
      throw new errors.UnableToReadKeystore();
    }
  }
  save() {
    const text = JSON.stringify(this.data);

    // The encrypted text and the version are colon delimited to make
    // it easy to visually read the version as we could have easily
    // included it with the buffer

    const keystore = [VERSION, Keystore.encrypt(text, this.password)].join(':');
    (0, _fs.writeFileSync)(this.path, keystore);
  }
  async load() {
    try {
      if (this.hasPassword() && !this.password) {
        if (process.env.KBN_KEYSTORE_PASSPHRASE_FILE) {
          this.password = (0, _fs.readFileSync)(process.env.KBN_KEYSTORE_PASSPHRASE_FILE, {
            encoding: 'utf8'
          }).trim();
        } else if (process.env.KEYSTORE_PASSWORD) {
          this.password = process.env.KEYSTORE_PASSWORD;
        } else {
          this.password = await (0, _prompt.question)('Enter password for the kibana keystore', {
            mask: '*'
          });
        }
      }
      const keystore = (0, _fs.readFileSync)(this.path);
      const [, data] = keystore.toString().split(':');
      this.data = JSON.parse(Keystore.decrypt(data, this.password));
    } catch (e) {
      if (e.code === 'ENOENT') {
        return;
      }
      throw e;
    }
  }
  reset() {
    this.data = {};
  }
  exists() {
    return (0, _fs.existsSync)(this.path);
  }
  keys() {
    return Object.keys(this.data);
  }
  has(key) {
    return this.keys().indexOf(key) > -1;
  }
  add(key, value) {
    this.data[key] = value;
  }
  remove(key) {
    delete this.data[key];
  }
  hasPassword() {
    try {
      const keystore = (0, _fs.readFileSync)(this.path);
      const [, data] = keystore.toString().split(':');
      Keystore.decrypt(data);
    } catch (e) {
      if (e instanceof errors.UnableToReadKeystore) {
        return true;
      }
    }
    return false;
  }
  setPassword(password) {
    this.password = password;
  }
}
exports.Keystore = Keystore;
(0, _defineProperty2.default)(Keystore, "errors", errors);
```

and as you see in the code snippet above its return that they're using `aes-256-gcm` to encrypt the keystore with `crypto` library and also save as a formatted like this

```js
 const text = JSON.stringify(this.data);

    // The encrypted text and the version are colon delimited to make
    // it easy to visually read the version as we could have easily
    // included it with the buffer

    const keystore = [VERSION, Keystore.encrypt(text, this.password)].join(':');
    (0, _fs.writeFileSync)(this.path, keystore);
```

here is example result of creating process
```
1:NiES2H4R9ksNcApbu6e3H7bCnNexrZD5BSg6xjdb9Jt0DYhq7zaJXmYSEuCTZPiNefiyva/78R1hW5vkGo8EPszkKUENpBBgkV2qx/O62DkEeVp9UGaO7N9YvOG3dg==
```

So this mean you have to extract information in .jks or .p12 formatted with tools like `keytool`
and import them mannually with encrypting procress. So imo it is the different method to storing the data between java keystore and kibina-keystore that using node.js until they could work with import module.

dgomezleon commented:
Hi @chocolattej 

Thanks a lot for that detailed explanation. Let's go ahead with your changes.
